js/src/redux/providers/tokensActions.js

```diff
@@ -14,56 +14,225 @@
+      // Check if we have data from the right contract
+      if (cached.tokenreg === tokenRegContract.address && cached.tokens) {
+        log.debug('found cached tokens', cached.tokens);
+        dispatch(_setTokens(cached.tokens));
+
+        // Fetch all the tokens images on load
+        // (it's the only thing that might have changed)
+        const tokenIndexes = Object.values(cached.tokens)
+          .filter((t) => t && t.fetched)
+          .map((t) => t.index);
+
+        fetchTokensData(tokenRegContract, tokenIndexes)(dispatch, getState);
+      } else {
+        store.remove(lsKey);
+      }
+    }
+  };
+}
+
 export function loadTokens (options = {}) {
   log.debug('loading tokens', Object.keys(options).length ? options : '');
 
   return (dispatch, getState) => {
     const { tokenReg } = Contracts.get();
 
-    tokenReg.getInstance()
-      .then((tokenRegInstance) => {
-        return fetchTokenIds(tokenRegInstance);
+    return tokenReg.getContract()
+      .then((tokenRegContract) => {
+        loadCachedTokens(tokenRegContract)(dispatch, getState);
+        return fetchTokenIds(tokenRegContract.instance);
       })
-      .then((tokenIndexes) => dispatch(fetchTokens(tokenIndexes, options)))
+      .then((tokenIndexes) => loadTokensBasics(tokenIndexes, options)(dispatch, getState))
       .catch((error) => {
         console.warn('tokens::loadTokens', error);
       });
   };
 }
 
+export function loadTokensBasics (_tokenIndexes, options) {
+  const limit = 64;
+
+  return (dispatch, getState) => {
+    const { api, tokens } = getState();
+    const { tokenReg } = Contracts.get();
+    const nextTokens = {};
+    const prevTokensIndexes = Object.values(tokens).map((t) => t.index);
+
+    // Only fetch tokens we don't have yet
+    const tokenIndexes = _tokenIndexes.filter((tokenIndex) => {
+      return !prevTokensIndexes.includes(tokenIndex);
+    });
+
+    const count = tokenIndexes.length;
+
+    log.debug('loading basic tokens', tokenIndexes);
+
+    if (count === 0) {
+      return Promise.resolve();
+    }
+
+    return tokenReg.getContract()
+      .then((tokenRegContract) => {
+        let promise = Promise.resolve();
+
+        for (let start = 0; start < count; start += limit) {
+          promise = promise
+            .then(() => fetchTokensBasics(api, tokenRegContract, start, limit))
+            .then((results) => {
+              results
+                .forEach((token) => {
+                  nextTokens[token.id] = token;
+                });
+            });
+        }
+
+        return promise;
+      })
+      .then(() => {
+        log.debug('fetched tokens basic info', nextTokens);
+
+        dispatch(setTokens(nextTokens));
+      })
+      .catch((error) => {
+        console.warn('tokens::fetchTokens', error);
+      });
+  };
+}
+
 export function fetchTokens (_tokenIndexes, options = {}) {
-  const tokenIndexes = uniq(_tokenIndexes || []);
+  const tokenIndexes = uniq(_tokenIndexes || []).sort();
```

Comment:
Why does it need to be sorted?

Reply:
I thought about doing this in a different way, which would have needed a sorted array.... Forgot to remove it!